03 Writing Algorithms/20 Historical Data/02 Warm Up Periods/99 Examples.html

```diff
@@ -1,4 +1,225 @@
+    {
+        SetStartDate(2021, 1, 1);
+        SetEndDate(2022, 1, 1);
+
+        // Request SPY data for signal generation and trading.
+        _spy = AddEquity("SPY", Resolution.Minute).Symbol;
+
+        // The candlestick patterns are based on daily consolidated trade bar.
+        var consolidator = new TradeBarConsolidator(TimeSpan.FromDays(1));
+        // Subscribe for automatically updating the indicators with the 1-day consolidator.
+        RegisterIndicator(_spy, _bbands, consolidator);
+        // Add event handler on candlestick indicators updated to trade the indicator.
+        _bbands.Updated += OnUpdated;
+
+        // Warm up the indicator and the SPY price data.
+        SetWarmUp(TimeSpan.FromDays(21));
+    }
+
+    private void OnUpdated(object sender, IndicatorDataPoint point)
+    {
+        // Trade according to the updated indicator values.
+        var upperBand = (sender as BollingerBands).UpperBand;
+        var lowerBand = (sender as BollingerBands).LowerBand;
+        var price = Securities[_spy].Price;
+
+        // Trade mean-reversal of the Bollinger Band.
+        if (price &gt; upperBand &amp;&amp; !Portfolio[_spy].IsShort)
+        {
+            SetHoldings(_spy, -0.5m);
+        }
+        else if (price &lt; lowerBand &amp;&amp; !Portfolio[_spy].IsLong)
+        {
+            SetHoldings(_spy, 0.5m);
+        }
+    }
+}</pre>
+    <pre class="python">class WarmUpPeriodsAlgorithm(QCAlgorithm):
+    bbands = BollingerBands(20, 2)
+
+    def initialize(self) -&gt; None:
+        self.set_start_date(2021, 1, 1)
+        self.set_end_date(2022, 1, 1)
+        
+        # Request SPY data for signal generation and trading.
+        self.spy = self.add_equity("SPY", Resolution.MINUTE).symbol
+
+        # The candlestick patterns are based on daily consolidated trade bar.
+        consolidator = TradeBarConsolidator(timedelta(1))
+        # Subscribe for automatically updating the indicators with the 1-day consolidator.
+        self.register_indicator(self.spy, self.bbands, consolidator)
+        # Add event handler on candlestick indicators updated to trade the indicator.
+        self.bbands.updated += self.on_updated
+
+        # Warm up the indicator and the SPY price data.
+        self.set_warm_up(timedelta(21))
+
+    def on_updated(self, sender: object, point: IndicatorDataPoint) -&gt; None:
+        # Trade according to the updated indicator values.
+        upper_band = sender.upper_band.current.value
+        lower_band = sender.lower_band.current.value
+        price = self.securities[self.spy].price
+
+        # Trade mean-reversal of the Bollinger Band.
+        if price &gt; upper_band and not self.portfolio[self.spy].is_short:
+            self.set_holdings(self.spy, -0.5)
+        if price &lt; lower_band and not self.portfolio[self.spy].is_long:
+            self.set_holdings(self.spy, 0.5)</pre>
+</div>
+
+<h4>Example 2: Warm Up Universe Selection</h4>
```

Comment:
The second example doesn't use warm-up.
During warm-up, we should update the EMA to avoid the history request. The EMA will be updated with the close of Monday (week_start), so we need 60 weeks to get 60 data points.
Or we don't do week_start, universe selection is called every day.